modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,377 @@
+
+    def __init__(
+        self,
+        func: Callable,
+        is_output: bool = False,
+        repartition_after: bool = False,
+        fan_out: bool = False,
+        pass_partition_id: bool = False,
+        reduce_fn: Optional[Callable] = None,
+    ):
+        self.func = func
+        self.is_output = is_output
+        self.oids = []
+        self.fan_out = fan_out
+        self.reptn = repartition_after
+        self.pass_num_ptn = pass_partition_id
+        self.operators = None
+        self.reduce_fn = reduce_fn
+
+    @classmethod
+    def print_options(cls):
+        """
+        Prints the kwargs that Query objects accept.
+        """
+        print(  # noqa: T201
+            """is_output : bool, default: False
+    Whether this query is an output query and should be passed both to the next query, and directly to postprocessing.
+repartition_after : bool, default: False
+    Whether to repartition after this query is computed. Currently, repartitioning is only supported if there is 1 partition prior to repartitioning.
+fan_out : bool, default: False
+    Whether to fan out this node. If True and only 1 partition is passed as input, the partition is replicated `num_partition` times, and the function is called on each. The `reduce_fn` must also be specified.
+pass_partition_id : bool, default: False
+    Whether to pass the numerical partition id to the query.
+reduce_fn : Callable
+    The reduce function to apply if `fan_out` is set to True. This takes the `num_partition` partitions that result from this query, and combines them into 1 partition.
+output_id : int, default: None
+    An id to assign to this node if it is an output."""
+        )
+
+
+class PandasQueryPipeline(object):
+    """
+    Internal representation of a query pipeline.
+
+    This object keeps track of the functions that compose to form a query pipeline.
+
+    Parameters
+    ----------
+    df : modin.pandas.Dataframe
+        The dataframe to perform this pipeline on.
+    axis : int or modin.core.dataframe.base.dataframe.utils.Axis
+        The axis along which to partition the dataframe for this pipeline.
+    num_partitions : int, default: 16
+        The number of partitions to maintain for the batched dataframe.
+    """
+
+    def __init__(self, df, axis: Union[int, Axis], num_partitions: int = 16):
+        if Engine.get() != "Ray" or (
+            not isinstance(df._query_compiler._modin_frame, PandasOnRayDataframe)
+        ):  # pragma: no cover
+            ErrorMessage.not_implemented(
+                "Batch Pipeline API is only implemented for Ray Engine."
+            )
+        self.df = df
+        self.axis = Axis(axis)
+        self.num_partitions = num_partitions
+        NPartitions.put(self.num_partitions)
+        NPartitions.get = lambda: self.num_partitions
+        self.outputs = []
+        self.nodes_list = []
+        self.node_to_id = None
+
+        self.dcq = PandasOnRayDataframeVirtualPartition.drain_call_queue
```

Comment:
This is a little hacky since we're modifying the `PandasOnRayDataframeVirtualPartition`'s method. We can instead pass `num_splits` as a kwarg and modify the `drain_call_queue` to accept num_splits - would that work better?